Validation tests for Reactive Messaging DeploymentExceptions

ryan-storey opened the pull request:
These tests check that DeploymentExceptions are correctly thrown as [per the spec](https://download.eclipse.org/microprofile/microprofile-reactive-messaging-3.0/microprofile-reactive-messaging-spec-3.0.html#_assembly_and_validation).

Closes #26274

abutch3r committed:
Override construction of DeploymentExceptions for Reactive Messaging startup

Override MediatorManager startup process to change the construction of the DeploymentException that is thrown to indicate wiring exceptions.

Update Reactive packages to latest
Mutiny - 2.2.0 -> 2.5.1
Reactive Messaging - 4.8.0 -> 4.10.1
Mutiny Convertor Api - 3.0.0 -> 3.0.1

Update EmitterImpl to match latest RM Provider version due to changes around blocking
Update WorkerPoolRegistry to include properties now used in MediatorManager

abutch3r committed:
Make requested updates

Include NLS bundle for our message for when an error occurs
Improve, building of errors string
Fix copyright

ryan-storey committed:
Add validation tests for DeploymentExceptions per spec

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/validation/IncomingConnectorMultipleDownstreams.java:
```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corporation and others.
```
is 2017 the right year here? or should it be 2023

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/validation/KafkaValidationTests.java:
```diff
@@ -0,0 +1,148 @@
+        AppValidator.validateAppOn(server)
+                        .withClass(ValidConfig.class)
+                        .withAppConfig(appConfig)
+                        .failsWith("Unknown connector for `" + ValidConfig.INCOMING_CHANNEL + "`.")
```
should try with a regex similar to `"SRMSG00072*" + ValidConfig.INCOMING_CHANNEL` that uses the Weld  Msg Code and the incoming channel in case the rest of the message content does change for some reason in the future

ryan-storey committed:
Rebased with additional error outputs and added extra output checks

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/kafka/validation/KafkaValidationTests.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Test
+    public void testMissingConnector() throws Exception {
```
At the end of each test you should `Mark` your place, this should prevent the chance of a false positive if another test causes the same error signature. This can be done with `server.setMarkToEndOfLog();`

abutch3r replied:
reviewing the AppValidator class, that does set the mark before it actually deploys the application. so you shouldn't need to set in the tests themselves

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/validation/AppValidator.java:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * Begin setting up an app validation to run on the given liberty server
+     */
+    public static AppValidator validateAppOn(LibertyServer server) {
```
One frustration I have with this class is that we effectively end up with an arbitary testApp name which can only be matched by a regex as you have done, but there is no way to guarantee within a test what that app name is and we are checking that in the logs. I would recommend added a property called appName that can be set on the AppValidator before `.run()` is called and uses that. you can then set them in the individual tests and then check for that explicit appName and that will reduce potential false-postivites. It should use the mechanism already used here if appName is null

ryan-storey committed:
Add appName attribute to AppValidator and address initial review
comments

Add appName attribute to AppValidator and address initial review
comments

Refactor AppValidator - move to FAT common bundle

Generalise AppValidator libraries and manifests configuration

Remove ReactiveMessagingExtension

Update AppValidator.java

ryan-storey commented:
#build

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/validation/AppValidator.java:
```diff
@@ -0,0 +1,180 @@
+                    app = app.addClass(clazz);
+                }
+                if (appConfig != null) {
+                    app = app.addAsLibraries(kafkaClientLibs())
```
I would have a generic methods on the class that could add libraries and manifests to the app, so that we don't have to include kafka specific entities within an application that might not use kakfka.

Also putting this in the appConfig block when they are unrelated doesn't make sense.

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/validation/AppValidator.java:
```diff
@@ -0,0 +1,180 @@
+ * This creates a web application containing just {@code MyInvalidClass}, deploys it, checks that it fails to start and that a message containing {@code CWMFT5001E} is found in the
+ * log and then undeploys it.
+ */
+public class AppValidator {
```
I would move based on later comments to either a `utils` package in this project or into the RM Common FAT project

ryan-storey commented:
#build

ryan-storey commented:
#build

abutch3r commented on dev/io.openliberty.microprofile.reactive.messaging.internal_fat.common/src/com/ibm/ws/microprofile/reactive/messaging/fat/AppValidator.java:
```diff
@@ -0,0 +1,202 @@
+ * </pre>
+ *
+ * <p>
+ * This creates a web application containing just {@code MyInvalidClass}, deploys it, checks that it fails to start and that a message containing {@code CWMFT5001E} is found in the
```
given this can do both successful and failure scenarios, I would correct the comment about it only being for failures

abutch3r reviewed (dismissed):
Only small bit about the AppValidator javadoc comments given that it can do either success or failure, but otherwise looks pretty much ready to go

abutch3r commented:
#build
#spawn.fullfat.butckets=io.openliberty.microprofile.reactive.messaging.internal_fat

abutch3r commented:
#build
#spawn.fullfat.buckets=io.openliberty.microprofile.reactive.messaging.internal_fat

abutch3r commented:
#libby

ryan-storey commented:
#build

abutch3r commented:
#libby

abutch3r commented:
CI Orchestrator failures are known issues with other those buckets and unrelated to the test changes here
